gems/aws-sdk-core/lib/aws-sdk-core/middleware/checksums.rb

```diff
@@ -0,0 +1,279 @@
+        # Handle request checksum
+        checksum_algorithm = input[@request_algorithm_member]
+
+        if checksum_algorithm
+          field_name = checksum_field_name(checksum_algorithm)
+          digest = AWS::SDK::Core::Checksums.algorithm_for(checksum_algorithm)
+
+          if @streaming
+            compute_streaming_checksum(context, digest, field_name)
+
+          else
+            context.request.headers[field_name] =
+              compute_checksum(digest, context.request.body)
+          end
+        end
+
+        # Handle response checksum
+        if @request_validation_mode_member &&
+           input[@request_validation_mode_member] == 'ENABLED'
+          # Compute an ordered list as the union between priority supported and
+          # the operation's modeled response algorithms.
+          validation_list = CHECKSUM_ALGORITHM_PRIORITIES & @response_algorithms
+
+          @response_validation_body = context.response.body =
+            ChecksumValidationIO.new(
+              context.response.body,
+              validation_list,
+              context
+            )
+        end
+
+        output = @app.call(input, context)
+        @response_validation_body&.validate!
+        output
+      end
+
+      private
+
+      def compute_streaming_checksum(context, digest, field_name)
+        headers = context.request.headers
+        headers['Content-Encoding'] = 'aws-chunked'
+        headers['X-Amz-Trailer'] = field_name
+        if @signed_streaming
+          headers['x-amz-content-sha256'] =
+            'STREAMING-AWS4-HMAC-SHA256-PAYLOAD-TRAILER'
+        else
+          headers['x-amz-content-sha256'] =
+            'STREAMING-UNSIGNED-PAYLOAD-TRAILER'
+        end
+
+        # some operations require the decoded content length set
+        # this requires that the original body has size
+        if @require_decoded_content_length
+          unless context.request.body.respond_to?(:size)
+            raise ArgumentError, 'Could not determine length of the body'
+          end
+
+          headers['X-Amz-Decoded-Content-Length'] =
+            context.request.body.size
+          context.request.body = AwsChunkedTrailerDigestSizeIO.new(
+            context.request.body, digest, field_name, @signed_streaming
+          )
+        else
+          # skip setting Content-Length
+          headers['transfer-encoding'] = 'chunked'
+          context.request.body = AwsChunkedTrailerDigestIO.new(
+            context.request.body, digest, field_name, @signed_streaming
+          )
+        end
+      end
+
+      def checksum_field_name(checksum_algorithm)
+        "x-amz-checksum-#{checksum_algorithm.downcase}"
+      end
+
+      def compute_checksum(digest, body)
+        if body.respond_to?(:read)
+          update_in_chunks(digest, body)
+        else
+          digest.update(body)
+        end
+        digest.base64digest
+      end
+
+      def update_in_chunks(digest, io)
```

Comment:
Should we use a wrapped body here?

Reply:
I don't think so - we need to compute the checksum upfront here, not when sending.